File: files/upload-xmlrpc.pl
#!/usr/bin/perl

# Use XML RPC to upload a file into a running exist and set permissions, or to
# execute an XQuery.
#
# Usage: upload-xmlrpc.pl [opts] dbpath < datafile
#        execute-xmlrpc.pl [opts] < xqueryfile
# Help:  upload-xmlrpc.pl -h

use Data::Dumper;
use File::Basename;
use Getopt::Std;
use RPC::XML::Client;
use strict;
use warnings;

# config defaults
my $CONF = {
    user   => 'admin',
    pass   => '',
    host   => 'localhost',
    port   => 8443,
    tls    => 1,
    mime   => 'application/xml',
    parse  => 1,
    replace => 1,
    owner  => 'admin',
    group  => 'SYSTEM',
    mode   => 'rw-r--r--',
    debug  => 0,
};

# check how this program was called
my ($PROGNAME) = fileparse($0);

# parse cmdline options, override defaults
my $OPTS = {};
getopts('dg:hH:Lm:M:o:p:P:Qu:T', $OPTS) or usage();
usage()                     if $OPTS->{h};
$CONF->{debug} = 1          if $OPTS->{d};
$CONF->{user}  = $OPTS->{u} if $OPTS->{u};
$CONF->{pass}  = $OPTS->{p} if $OPTS->{p};
$CONF->{host}  = $OPTS->{H} if $OPTS->{H};
$CONF->{port}  = $OPTS->{P} if $OPTS->{P};
$CONF->{tls}   = 0          if $OPTS->{T};
$CONF->{parse} = 0          if ($OPTS->{Q} or $OPTS->{L});
$CONF->{mime}  = 'application/xquery' if $OPTS->{Q};
$CONF->{mime}  = 'text/html'          if $OPTS->{L};
$CONF->{mime}  = $OPTS->{M} if $OPTS->{M};
$CONF->{owner} = $OPTS->{o} if $OPTS->{o};
$CONF->{group} = $OPTS->{g} if $OPTS->{g};
$CONF->{mode}  = $OPTS->{m} if $OPTS->{m};

# read filename from args and content from STDIN
my $fname  = shift @ARGV;
my @data   = <STDIN>;
my $data   = join("", @data);

# connect to exist XML RPC
my $url    = sprintf("%s://%s:%s\@%s:%d/exist/xmlrpc",
		     $CONF->{tls} ? 'https' : 'http',
		     $CONF->{user}, $CONF->{pass},
		     $CONF->{host}, $CONF->{port});
my $client = RPC::XML::Client->new($url);
#my $client = RPC::XML::Client->new($url, 'useragent', ['ssl_opts', { verify_hostname => 0 }]);

# execute query instead of uploading a file
if ($PROGNAME eq 'execute-xmlrpc.pl') {
    my $code   = RPC::XML::string->new($data);
    my $num    = RPC::XML::int->new(100);
    my $start  = RPC::XML::int->new(0);
    my $parms  = RPC::XML::struct->new({});
    my $exreq  = RPC::XML::request->new('query', $code, $num, $start, $parms);
    my $exres  = $client->send_request($exreq);
    my $exresp;
    if (ref $exres) {
	$exresp = $exres->value;
	print "Execute response: $exresp\n" if $CONF->{debug};
	exit;
    } else {
	print STDERR "Error executing XQuery: $exres\n";
	exit 4;
    }
}

# upload file
my $b64    = RPC::XML::base64->new($data);
my $upreq  = RPC::XML::request->new('upload', $b64, length $b64->value());
my $upres  = $client->send_request($upreq);
my $upresp;
if (ref $upres) {
    $upresp = $upres->value;
    print "Upload response: $upresp\n" if $CONF->{debug};
} else {
    print STDERR "Error uploading file: $upres\n";
    exit 1;
}

# call parseLocal() to store, need to force some explicit data types here
my $desc    = RPC::XML::string->new($upres);
my $parse   = RPC::XML::boolean->new($CONF->{parse});
my $replace = RPC::XML::boolean->new($CONF->{replace});
my $pareq   = RPC::XML::request->new('parseLocalExt', $desc, $fname, 
				     $replace, $CONF->{mime}, $parse);
my $pares   = $client->send_request($pareq);
my $paresp;
if (ref $pares) {
    $paresp = $pares->value;
    print "Store response: $paresp\n" if $CONF->{debug};
    print Dumper($paresp) if $CONF->{debug};
} else {
    print STDERR "Error storing file: $pares\n";
    exit 2;
}

# set owner/group/perms - perms are expected as "rw-r--r--", not 0644
my $mode    = RPC::XML::string->new($CONF->{mode});
my $spreq   = RPC::XML::request->new('setPermissions', $fname,
				     $CONF->{owner}, $CONF->{group}, $mode);
my $spres   = $client->send_request($spreq);
my $spresp;
if (ref $spres) {
    $spresp = $spres->value;
    print "SetPermissions response: $spresp\n" if $CONF->{debug};
    print Dumper($spresp) if $CONF->{debug};
} else {
    print STDERR "Error storing file: $spres\n";
    exit 3;
}

exit;


sub usage {
    print <<USAGE;
Use XML RPC to upload a file into a running exist and set permissions, or to
execute an XQuery.

Usage:

  upload-xmlrpc.pl [OPTIONS] target-filename
  execute-xmlrpc.pl [OPTIONS]

Filename is the eXist DB path where to store the file.
File content to upload or XQuery to execute is expected on STDIN.  Examples:

  upload-xmlrpc.pl /db/apps/foo/bar.xml < bar.xml
  execute-xmlrpc.pl < foo.xql

OPTIONS:

-d       enable debug output
-h       help (this text)
-u user  username to login to eXist XML RPC (default: admin)
-p pass  password to login to eXist XML RPC (default: "")
-H host  URL hostname (default: localhost)
-P port  URL port (default: 8443)
-T       disable TLS connection (default: enabled)
-o owner set owner of uploaded file (default: admin)
-g group set group of uploaded file (default: SYSTEM)
-m mode  set mode  of uploaded file (default: rw-r--r--) - NOT 0644 format!
-M type  set MIME type of uploaded file (default: application/xml)
-Q       uploaded file is XQuery, shortcut for "-M application/xquery"
-L       uploaded file is HTML, shortcut for "-M text/html"
-B       uploaded file is binary [NOT IMPLEMENTED YET]

You probably need to use "-T -P 8080" to force unencrypted HTTP connections
to localhost if untrusted X.509 certs are used.

USAGE
    exit 1;
}

File: files/upload-xmlrpc.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import argparse
import base64
import os
import socket
import sys

from xmlrpc.client import ServerProxy, Error

parser = argparse.ArgumentParser()
#parser.add_argument("echo", help="echo the string you use here", type=str)
#parser.add_argument("num", type=int, help="double the number you use here")
parser.add_argument("-d", "--debug",  action="store_true", help="enable debug output")
parser.add_argument("-t", "--timeout", type=int, default=180, help="connection timeout")
parser.add_argument("-u", "--user", default="admin", help="username to login to eXist XML RPC (default: admin)")
parser.add_argument("-p", "--pass", default="", help="password to login to eXist XML RPC (default: \"\")")
parser.add_argument("-H", "--host", default="localhost", help="URL hostname (default: localhost)")
parser.add_argument("-P", "--port", type=int, default=8080, help="URL port (default: 8443)")
parser.add_argument("-T", action="store_true", help="disable TLS connection (default: enabled)")
parser.add_argument("-o", "--owner", default="admin", help="set owner of uploaded file (default: admin)")
parser.add_argument("-g", "--group", default="SYSTEM", help="set group of uploaded file (default: SYSTEM)")
parser.add_argument("-m", "--mode", default="rw-r--r--", help="set mode  of uploaded file (default: rw-r--r--) - NOT 0644 format!")
parser.add_argument("-M", "--mime", default="application/xml", help="set MIME type of uploaded file (default: application/xml)")
parser.add_argument("-Q", action="store_true", help="uploaded file is XQuery, shortcut for \"-M application/xquery\"")
parser.add_argument("-L", action="store_true", help="uploaded file is HTML, shortcut for \"-M text/html\"")
parser.add_argument("-B", action="store_true", help="uploaded file is binary [NOT IMPLEMENTED YET]")
parser.add_argument("file-name", default="", nargs="?", help="name of file to upload")
args = parser.parse_args()
#print(args)

xmlrpcDebug = args.debug
xmlrpcTimeout = args.timeout
xmlrpcUser = args.user
xmlrpcPass = vars(args)['pass']
xmlrpcHost = args.host
xmlrpcPort = args.port
xmlrpcSchema = "https"
if args.T:
    xmlrpcSchema = "http"
xmlrpcOwner = args.owner
xmlrpcMode = args.mode
xmlrpcMime = args.mime
if args.Q:
    xmlrpcMime = "application/xquery"
if args.L:
    xmlrpcMime = "text/html"
if args.B:
    xmlrpcMime = "application/octet-stream"

scriptName = os.path.basename(sys.argv[0])
#print(scriptName)
fname = sys.argv[-1]
#print(fname)

data = sys.stdin.read().encode()
#data = "\n".join(sys.stdin.readlines())
#print(data)
#sys.exit()



socket.setdefaulttimeout(xmlrpcTimeout)

with ServerProxy('{xmlrpcSchema}://{xmlrpcUser}:{xmlrpcPass}@{xmlrpcHost}:{xmlrpcPort}/exist/xmlrpc'.format(**locals())) as proxy:
    #print(proxy)
    if (scriptName == 'execute-xmlrpc.py'):
        print(scriptName)
        #xquery = '(<test1/>,<test2/>)'
        xquery = data
        limitResultNumberTo = 100 # 0 to disable
        startWithResultNumber = 1
        params = {}
        try:
            print(proxy.query(xquery, limitResultNumberTo, startWithResultNumber, params))
        except Error as v:
            print("ERROR", v)
            sys.exit(-1)
        pass
        sys.exit()
    elif (scriptName == 'upload-xmlrpc.py'):
        #print(scriptName)
        #b64 = base64.b64encode(data.encode('utf-8'))
        upres = -1
        pares = -1
        spres = -1
        try:
            #upres = proxy.upload(data, len(base64(data)))
            upres = proxy.upload(data, len(data))
            print(upres)
            #pares = proxy.parseLocalExt(str(upres), fname, replace=1, mime='application/xml', parse=1)
            pares = proxy.parseLocalExt(str(upres), fname, 1, xmlrpcMime, 1)
            print(pares)
            #spres = proxy.setPermissions(fname, owner='admin', group='SYSTEM', mode='rw-r--r--')
            spres = proxy.setPermissions(fname, 'admin', 'SYSTEM', 'rw-r--r--')
            print(spres)
        except Error as v:
            print("ERROR", v)
            sys.exit(-1)
        pass
        sys.exit()

if __name__=='__main__':
    #print("hello, world!")
    pass
